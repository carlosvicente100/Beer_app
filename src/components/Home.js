import React, { useState } from 'react'
import styled from 'styled-components'
import { useSelector } from 'react-redux'
import { filterByDate, makePagination } from '../utils'

import Filters from './Filters'
import ItemList from './ItemList'
import Pagination from './Pagination'

const HomeContainer = styled.div`
  font-family: verdana;
  height: -webkit-fill-available;
  background-color: #2c2c54;
  display: grid;
  grid-template-columns: auto 500px auto;
  row-gap: 20px;
  grid-template-areas:
    ". Filters ."
    "ItemList ItemList ItemList"
    "NoResults NoResults NoResults"
    ". Pagination .";

  div.Filters {
    grid-area: Filters;
  }
  div.ItemList {
    grid-area: ItemList;
  }
  div.NoResults {
    background:#2980b9;
    color:white;
    display: flex;
    justify-content: center;
    align-items: center;
    width 100%;
    height:300px;
    grid-area: NoResults;
  }
  div.Pagination {
    grid-area: Pagination;
  }
`
const Home = () => {
  const [currentPage, setCurrentPage] = useState(0)
  const { items = [], error } = useSelector((state) => state.itemsReducer)
  const { startDate, endDate } = useSelector((state) => state.filtersReducer)
  let itemsFiltered = []
  if (items.length > 0) {
    itemsFiltered = filterByDate(startDate, endDate, items)
  }
  let paginatedItems = makePagination(itemsFiltered)

  const updateFilters = () => {
    setCurrentPage(0)
  }

  return (
    <HomeContainer className="homecontainer">
      <Filters updateFilters={updateFilters}></Filters>
      {paginatedItems.length > 0 ? (
        <>
          <ItemList items={paginatedItems[currentPage]}></ItemList>
          <Pagination
            setCurrentPage={setCurrentPage}
            totalPagination={paginatedItems}
            current={currentPage}
          ></Pagination>
        </>
      ) : (
        <div className="NoResults">
          <p>{error ? "We're having problems getting Beers, try it in a few minutes" : 'no beer for you :( '}</p>
        </div>
      )}
    </HomeContainer>
  )
}

export default Home
