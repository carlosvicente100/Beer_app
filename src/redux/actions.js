export const fetchRequested = "FETCH_REQUESTED";
export const fetchRequestedItems = "FETCH_REQUESTED_ITEMS";
export const fetchSuccessItems = "FETCH_SUCCESS_ITEMS";
export const fetchErrorItems = "FETCH_ERROR_ITEMS";

//filters
export const fetchSetDates = "FETCH_SET_DATES";
export const fetchResetDates = "FETCH_RESET_DATES";
