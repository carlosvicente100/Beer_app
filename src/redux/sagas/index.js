import { all } from 'redux-saga/effects'
import fetchDataFromUrl from './fetch'

export default function* rootSaga() {
  yield all([fetchDataFromUrl()])
}
