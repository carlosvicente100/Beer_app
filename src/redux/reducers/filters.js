import { fetchSetDates, fetchResetDates } from "../actions";

export const initialState = {
  startDate: {
    year: -1,
    month: -1
  },
  endDate: {
    year: -1,
    month: -1
  }
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSetDates:
      return {
        ...state,
        startDate: action.data.startDate,
        endDate: action.data.endDate
      };
    case fetchResetDates:
      return {
        ...initialState
      };
    default:
      return {
        ...state
      };
  }
}
