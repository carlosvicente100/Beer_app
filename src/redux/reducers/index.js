import { combineReducers } from 'redux'

import itemsReducer from './items'
import filtersReducer from './filters'

export default combineReducers({
  itemsReducer,
  filtersReducer
})
