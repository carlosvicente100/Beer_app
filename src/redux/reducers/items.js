import { fetchSuccessItems, fetchErrorItems } from '../actions'

export const initialState = {
  error: null,
  items: []
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSuccessItems:
      return {
        ...state,
        items: action.data
      }
    case fetchErrorItems:
      return {
        ...state,
        error: true
      }
    default:
      return {
        ...state
      }
  }
}
