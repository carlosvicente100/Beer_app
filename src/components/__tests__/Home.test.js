import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { render as rtlRender } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { initialState as initialStateFilters } from '../../redux/reducers/filters'
import { initialState as initialStateItems } from '../../redux/reducers/items'
import reducer from '../../redux/reducers'

import itemListFixture from '../../fixtures/itemList'

import Home from '../Home'

beforeEach(() => {
  jest.clearAllMocks()
})

const initialReducerState = {
  filtersReducer: initialStateFilters,
  itemsReducer: initialStateItems
}

function render(ui, { initialState, store = createStore(reducer, initialState), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

describe('Home Component Suite', () => {
  test('Snapshot check', () => {
    expect(render(<Home />, { initialReducerState }).baseElement).toMatchSnapshot()
  })

  describe('Home Structure', () => {
    test('Should have Filters, ItemList and Pagination - with items to list', () => {
      const store = createStore(() => ({
        filtersReducer: initialStateFilters,
        itemsReducer: {
          error: null,
          items: itemListFixture
        }
      }))
      const { container } = render(<Home />, { store })
      expect(container.getElementsByClassName('Filters').length).toBe(1)
      expect(container.getElementsByClassName('ItemList').length).toBe(1)
      expect(container.getElementsByClassName('Pagination').length).toBe(1)
      expect(container.getElementsByClassName('NoResults').length).toBe(0)
    })
    test('Should have Filters, NoResults and Pagination - without items to list', () => {
      const { container } = render(<Home />)
      expect(container.getElementsByClassName('Filters').length).toBe(1)
      expect(container.getElementsByClassName('NoResults').length).toBe(1)
      expect(container.getElementsByClassName('ItemList').length).toBe(0)
      expect(container.getElementsByClassName('Pagination').length).toBe(0)
    })
  })
})
