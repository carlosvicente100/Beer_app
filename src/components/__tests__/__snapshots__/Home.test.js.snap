// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Suite Snapshot check 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c4 {
  margin: auto;
}

.c1 {
  background-color: #227093;
  color: white;
  text-align: center;
}

.c2 {
  width: 50%;
  display: inline-grid;
  grid-area: input;
  grid-template-areas: 'span' 'select';
}

.c2 span {
  grid-area: span;
}

.c2 div.Select {
  width: 100%;
  grid-area: select;
}

.c5 {
  background-color: #2980b9;
  border: none;
  color: white;
  padding: 12px 16px;
  font-size: 16px;
  cursor: pointer;
}

.c5:hover {
  background-color: tomato;
}

.c5:disabled {
  background-color: #7f8c8d;
}

.c0 {
  font-family: verdana;
  height: -webkit-fill-available;
  background-color: #2c2c54;
  display: grid;
  grid-template-columns: auto 500px auto;
  row-gap: 20px;
  grid-template-areas: ". Filters ." "ItemList ItemList ItemList" "NoResults NoResults NoResults" ". Pagination .";
}

.c0 div.Filters {
  grid-area: Filters;
}

.c0 div.ItemList {
  grid-area: ItemList;
}

.c0 div.NoResults {
  background: #2980b9;
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 300px;
  grid-area: NoResults;
}

.c0 div.Pagination {
  grid-area: Pagination;
}

<body>
  <div>
    <div
      class="c0 homecontainer"
    >
      <div
        class="c1 Filters"
      >
        <label
          class="c2"
        >
          Start Month
          <div
            class="c3 Select"
          >
            <select
              class="c4"
              data-testid="select"
            >
              <option
                value="-1"
              >
                Empty
              </option>
              <option
                value="1"
              >
                January
              </option>
              <option
                value="2"
              >
                February
              </option>
              <option
                value="3"
              >
                March
              </option>
              <option
                value="4"
              >
                April
              </option>
              <option
                value="5"
              >
                May
              </option>
              <option
                value="6"
              >
                June
              </option>
              <option
                value="7"
              >
                July
              </option>
              <option
                value="8"
              >
                August
              </option>
              <option
                value="9"
              >
                September
              </option>
              <option
                value="10"
              >
                October
              </option>
              <option
                value="11"
              >
                November
              </option>
              <option
                value="12"
              >
                December
              </option>
            </select>
          </div>
        </label>
        <label
          class="c2"
        >
          End Month
          <div
            class="c3 Select"
          >
            <select
              class="c4"
              data-testid="select"
            >
              <option
                value="-1"
              >
                Empty
              </option>
              <option
                value="1"
              >
                January
              </option>
              <option
                value="2"
              >
                February
              </option>
              <option
                value="3"
              >
                March
              </option>
              <option
                value="4"
              >
                April
              </option>
              <option
                value="5"
              >
                May
              </option>
              <option
                value="6"
              >
                June
              </option>
              <option
                value="7"
              >
                July
              </option>
              <option
                value="8"
              >
                August
              </option>
              <option
                value="9"
              >
                September
              </option>
              <option
                value="10"
              >
                October
              </option>
              <option
                value="11"
              >
                November
              </option>
              <option
                value="12"
              >
                December
              </option>
            </select>
          </div>
        </label>
        <label
          class="c2"
        >
          Start year
          <div
            class="c3 Select"
          >
            <select
              class="c4"
              data-testid="select"
            >
              <option
                value="-1"
              >
                Empty
              </option>
            </select>
          </div>
        </label>
        <label
          class="c2"
        >
          End year
          <div
            class="c3 Select"
          >
            <select
              class="c4"
              data-testid="select"
            >
              <option
                value="-1"
              >
                Empty
              </option>
            </select>
          </div>
        </label>
        <button
          class="c5"
          disabled=""
        >
          Filter
        </button>
        <button
          class="c5"
          disabled=""
        >
          Reset All
        </button>
      </div>
      <div
        class="NoResults"
      >
        <p>
          no beer for you :( 
        </p>
      </div>
    </div>
  </div>
</body>
`;
